### PR TITLE
Dynamic transaction description generation

### PR DESCRIPTION
**Pull request by znarf**
### Samples

Collective context

```
"description"
"Monthly contribution from Marinus Angelo Klap"
"Host Fee to Open Source Collective"
"Monthly contribution from coneyware (Backers)"
"Host Fee to Open Source Collective"
"Monthly contribution from Ayush Sachdeva (Backers)"
"Host Fee to Open Source Collective"
"Monthly contribution from Masahiro Takeda"
"Host Fee to Open Source Collective"
"Monthly contribution from Bryan Huizi"
"Host Fee to Open Source Collective"
"Monthly contribution from Pavel Shermayster (Backers)"
"Host Fee to Open Source Collective"
"Expense from Henry Zhu - Babel Maintenance (August 2021)"
"Expense from Nicolò Ribaudo - Babel development and support (August 2021)"
"Expense from Jùnliàng Huáng - Babel Engineering Support (August 2021)"
```

Collective context with full description

```
"description"
"Monthly contribution from Marinus Angelo Klap to Babel"
"Host Fee from Babel to Open Source Collective"
"Monthly contribution from coneyware to Babel (Backers)"
"Host Fee from Babel to Open Source Collective"
"Monthly contribution from Ayush Sachdeva to Babel (Backers)"
"Host Fee from Babel to Open Source Collective"
"Monthly contribution from Masahiro Takeda to Babel"
"Host Fee from Babel to Open Source Collective"
"Monthly contribution from Bryan Huizi to Babel"
"Host Fee from Babel to Open Source Collective"
"Monthly contribution from Pavel Shermayster to Babel (Backers)"
"Host Fee from Babel to Open Source Collective"
"Expense from Henry Zhu to Babel - Babel Maintenance (August 2021)"
"Expense from Nicolò Ribaudo to Babel - Babel development and support (August 2021)"
"Expense from Jùnliàng Huáng to Babel - Babel Engineering Support (August 2021)"
```

Host budget context

```
"description"
"Host Fee from FlyByWire Simulations"
"Host Fee Share to Open Collective"
"Host Fee Share Debt owed to Open Collective"
```

Host budget context with full description

```
"description"
"Host Fee from FlyByWire Simulations to Open Source Collective"
"Host Fee Share from Open Source Collective to Open Collective"
"Host Fee Share Debt owed to Open Collective by Open Source Collective"
```

Host transactions context

```
"type","description"
"CREDIT","Contribution from PostHog to Jest"
"CREDIT","Host Fee from Jest to Open Source Collective"
"DEBIT","Host Fee Share from Open Source Collective to Open Collective"
"DEBIT","Host Fee from Jest to Open Source Collective"
"CREDIT","Contribution from Guest to FlyByWire Simulations"
"CREDIT","Host Fee from FlyByWire Simulations to Open Source Collective"
"DEBIT","Host Fee Share from Open Source Collective to Open Collective"
"CREDIT","Host Fee Share Debt owed to Open Collective by Open Source Collective"
```

### CODE REVIEW

File: server/graphql/loaders/index.js [hunk 1]
```diff
@@ -565,6 +565,7 @@ export const loaders = req => {
 
   /** *** Transaction *****/
   context.loaders.Transaction = {
+    ...context.loaders.Transaction,
```
Comment (znarf) [inline server/graphql/loaders/index.js:568]: We were removing the default loaders ...

File: server/lib/transactions.js [hunk 2]
```diff
@@ -242,3 +247,126 @@ export async function sum(where) {
...
+    fromString = '',
+    toString = '';
+
+  if (transaction.isRefund && transaction.RefundTransactionId) {
+    const refundedTransaction = await (req
+      ? req.loaders.Transaction.byId.load(transaction.RefundTransactionId)
+      : models.Transactions.findByPk(order.RefundTransactionId));
+    if (refundedTransaction) {
+      const refundedTransactionDescription = await generateDescription(refundedTransaction, { req, full });
+      return `Refund of \`${refundedTransactionDescription}\``;
```
Comment (kewitz) [inline server/lib/transactions.js:279]: Do we need to use backticks here? 
```suggestion
      return `Refund of "${refundedTransactionDescription}"`;
```

File: server/lib/transactions.js [hunk 3]
```diff
@@ -242,3 +247,126 @@ export async function sum(where) {
...
+    }
+  }
+
+  let order, expense, subscription, tier;
```
Comment (kewitz) [inline server/lib/transactions.js:283]: These variables are never reassigned, I would rather use `const` and assign them whenever needed.

File: server/lib/transactions.js [hunk 4]
```diff
@@ -242,3 +247,126 @@ export async function sum(where) {
...
+  HOST_FEE: `Host Fee`,
+  HOST_FEE_SHARE: `Host Fee Share`,
+  HOST_FEE_SHARE_DEBT: `Host Fee Share Debt`,
+  PAYMENT_PROCESSOR_COVER: `Cover of Payment Processor Fee`,
+  PLATFORM_TIP: `Platform Tip`,
+  PLATFORM_TIP_DEBT: `Platform Tip Debt`,
+  PREPAID_PAYMENT_METHOD: `Prepaid Payment Method`,
+};
+
+export async function generateDescription(transaction, { req = null, full = false } = {}) {
```
Comment (kewitz) [inline server/lib/transactions.js:265]: Nice signature! I like `req` is inside the options parameter.

File: server/lib/transactions.js [hunk 5]
```diff
@@ -242,3 +247,126 @@ export async function sum(where) {
...
+    if (subscription && subscription.interval === 'month') {
+      baseString = `Monthly contribution`;
+    } else if (subscription && subscription.interval === 'year') {
+      baseString = `Yearly contribution`;
+    } else if (tier && tier.type === TICKET) {
+      baseString = `Registration`;
+    }
+    if (tier) {
+      tierString = ` (${tier.name})`;
+    }
```
Comment (kewitz) [inline server/lib/transactions.js:313]:
```suggestion
    if (order?.TierId) {
      const tier = await (req ? req.loaders.Tier.byId.load(order.TierId) : models.Tiers.findByPk(order.TierId));
      tierString = ` (${tier.name})`;
    }
```

**Review by kewitz — APPROVED** [review]
Looks good! :shipit:

File: server/lib/transactions.js (hunk 2 above)
Comment (znarf) [reply server/lib/transactions.js:279]: Good question, my experience is that double quotes are sometime to be expected (in an Expense description for example, `SELECT * FROM "Transactions"  WHERE "description" ILIKE '%"%' AND "isRefund" IS FALSE`) and I was trying to find something that was not double quotes or single quotes.

Even if it's complicated, perhaps we could use backticks only if single quotes and double quotes are already used, JavaScript style. Or use a different syntax without quotes, "Refund: xxxx' or "Refund - xxxxx".

File: server/lib/transactions.js [hunk 6]
```diff
@@ -242,3 +247,126 @@ export async function sum(where) {
...
+    if (subscription && subscription.interval === 'month') {
+      baseString = `Monthly contribution`;
+    } else if (subscription && subscription.interval === 'year') {
+      baseString = `Yearly contribution`;
+    } else if (tier && tier.type === TICKET) {
+      baseString = `Registration`;
+    }
+    if (tier) {
+      tierString = ` (${tier.name})`;
```
Comment (Betree) [inline server/lib/transactions.js:312]: I would truncate `tier.name`

File: server/lib/transactions.js [hunk 7]
```diff
@@ -242,3 +247,126 @@ export async function sum(where) {
...
+    : models.Collectives.findByPk(order.CollectiveId));
+  const oppositeAccount = await (req
+    ? req.loaders.Collective.byId.load(transaction.FromCollectiveId)
+    : models.Collectives.findByPk(order.FromCollectiveId));
+
+  if (transaction.isDebt) {
+    debtString = ' owed';
+    if (transaction.type === CREDIT) {
+      if (full) {
+        toString = ` by ${account.name.trim()}`;
```
Comment (Betree) [inline server/lib/transactions.js:336]: `name` is supposed to be [trimmed](https://github.com/opencollective/opencollective-api/blob/776010bc0b0b39302778933d6269c22a445288a0/server/models/Collective.js#L161) on the model, is it not always the case? Maybe we can rather migrate if there are past entries that we want to fix.

File: server/lib/transactions.js (hunk 3 above)
Comment (znarf) [reply server/lib/transactions.js:283]: Uhm, that's not possible without code change regarding variable scopes and re-assigment.

File: server/lib/transactions.js (hunk 5 above)
Comment (znarf) [reply server/lib/transactions.js:313]: to be on the defensive side, it's possible that `tier` is null. If the tier was deleted.

File: server/lib/transactions.js (hunk 7 above)
Comment (znarf) [reply server/lib/transactions.js:336]: I saw some occurrences of space at the end.

**Pull request merged**